_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,233 @@
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: disabled # states that the cluster is ZooKeeper-based
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+The deployed cluster has severals pods running for the Kafka brokers and ZooKeeper nodes.
+
+```shell
+NAME                                          READY   STATUS    RESTARTS   AGE
+my-cluster-entity-operator-555785d699-kksrs   2/2     Running   0          17m
+my-cluster-kafka-0                            1/1     Running   0          18m
+my-cluster-kafka-1                            1/1     Running   0          18m
+my-cluster-kafka-2                            1/1     Running   0          18m
+my-cluster-zookeeper-0                        1/1     Running   0          20m
+my-cluster-zookeeper-1                        1/1     Running   0          20m
+my-cluster-zookeeper-2                        1/1     Running   0          20m
+strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
+```
+
+#### From ZooKeeper to full KRaft based brokers
+
+The first step is to deploy a `KafkaNodePool` custom resource which hosts the KRaft controllers and then changing the `strimzi.io/kraft` annotation from `disabled` to `migration` to trigger the process.
+
+Following a snippet of a `KafkaNodePool` custom resource to be deployed for getting the KRaft controllers running.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - controller
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: false
+```
+
+The corresponding `Kafka` custom resource is then updated in regards to the `strimzi.io/kraft` set to `migration`. 
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+The annotation change triggers the Strimzi operator to run the following steps for you:
+
+* deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled;
+* configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them;
+* checking the status of the migration, on each reconciliation, and updating the `Kafka` custom resource to provide such status to the user through the `status.kafkaMetadataState` field;
+* when the migration is done, the brokers are reconfigured to be not connected to ZooKeeper anymore, with the migration flag disabled and they are rolled again.
```

Comment:
```suggestion
* deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled
* configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them
* on each reconciliation, check the status of the migration and update the `Kafka` custom resource to provide such status to the user through`status.kafkaMetadataState`
* when the migration is done, reconfigure brokers to disconnect from ZooKeeper, with the migration flag disabled, and roll them again
```

Reply:
Still I used part of the suggestions only.